### ISSUE TITLE
Keyboard shortcuts should be configurable

### ISSUE THREAD

**Issue by szhorvat** [nteract/archived-desktop-app#5092]
**Is your feature request related to a problem? Please describe.**

The auto-completion shortcut, Ctrl-Space, conflicts with a system shortcut (trigger Spotlight on macOS)

**Describe the solution you'd like**

Make shortcuts configurable in nteract desktop.

**Describe alternatives you've considered**

A workaround is to change the system-wide shortcut to free up Ctrl-Space for nteract. I am not willing to do this because my setup work well, and nteract is _the only_ application that had this problem. All other programs that had Ctrl-Space set for auto-complete allowed it to be reconfigured.

**Additional context**

The lack of auto-completion makes nteract basically unusable.

**Comment by captainsafia** [comment]
> The auto-completion shortcut, Ctrl-Space, conflicts with a system shortcut (trigger Spotlight on macOS)

The default keyboard shortcut for Spotlight is <kbd>Cmd + Space</kbd>, not <kbd>Ctrl + Space</kbd> so there shouldn't be any conflict here.

> Make shortcuts configurable in nteract desktop.

There's a few different issues tracking this and we are currently tracking this work under https://github.com/nteract/nteract/issues/4781.

Keyboard shortcuts like the auto-complete shortcut can be configuring via CodeMirror. We currently disable in order to support registering our own event handlers for different things. It might be worthwhile to explore what allowing users to configure the `extraKeys` CodeMirror property via our existing CodeMirror configuration options.

**Comment by itko** [comment]
> tracking this work under #4781.

The issue you mention deals with consolidating the shortcut methods. It doesn't specifically address the inability to configure the keyboard shortcuts.

> Keyboard shortcuts like the auto-complete shortcut can be configuring via CodeMirror.

Can you you clarify what you mean here? Are you saying there is a way to configure the nteract shortcuts?

**Comment by captainsafia** [comment]
> Can you you clarify what you mean here? Are you saying there is a way to configure the nteract shortcuts?

CodeMirror's keyboard shortcuts are configured independently of nteract's app level keyboard shortcuts. These can be set via the `codemirror` config option in the config JSON. However, we currently disable this to avoid users overriding some of our default keymap configurations (e.g. requesting completions from the kernel when the user presses <kbd>Ctrl+.</kbd>).

It's possible for us to remove this constraint so that users can modify this.

> The issue you mention deals with consolidating the shortcut methods. It doesn't specifically address the inability to configure the keyboard shortcuts.

I'll update the issue comment to account that consolidation will make it easier to implement a config layer.

**Comment by CrystallineCat** [comment]
In addition to CodeMirror's, the keyboard shortcuts for all commands in the menu should already be configurable (introduced in #5137), using e.g.:
```json
{
  "keyboardShortcuts": {
    "PublishGist": "Ctrl+P",
    "ClearAll": "Ctrl+Alt+C"
  }
}
```

See https://github.com/nteract/nteract/blob/master/applications/desktop/src/common/accelerators.ts for a list of all default shortcuts and https://github.com/nteract/nteract/blob/master/applications/desktop/src/common/menu.ts for the names of the commands (in the form `commands.<Name>`, the `<Name>` part is the key to use in the config file).

**Comment by szhorvat** [comment]
Could you clarify:

 - Which file one needs to edit to change shortcuts?
 - Can the shortcut for auto-completion be changed? I do not see it in the files linked above.

**Comment by CrystallineCat** [comment]
The file is `nteract.json` in the Jupyter config dir (`~/.jupyter` on my machine). However, auto-complete can't be changed right now via the config. The only way to change it right now is to edit this line: https://github.com/nteract/nteract/blob/e81cbb2565be0b08e222a8c172e8316ae36e2f37/packages/editor/src/index.tsx#L147 
...and then build nteract yourself (`yarn build:all:cleanly && yarn dist`, the result is in `applications/desktop/dist`). :crying_cat_face:

**Comment by captainsafia** [comment]
@CrystallineCat I must've completely missed it in my review because I didn't realize that PR added support for customizing app-level keyboard shortcuts for the config file. We definitely need to let users know about this since it is a pretty highly requested feature.

I can update the changelog for the June release to reflect this and we'll need to add some documentation on docs.nteract.io about this as well. Can you submit a PR with docs on what keyboard shortcuts are configurable and how to modify them?

**Comment by szhorvat** [comment]
@captainsafia You are right that the default for Spotlight *in current macOS versions* is Cmd-Space. But I've been using certain global shortcuts for a long time, from before when that shortcut was added. Not using Ctrl-Space would be a major inconvenience for me (and I don't even know what other single-handed shortcut I could find that doesn't conflict with a lot of programs I already use), so I ended up simply not using nteract because of this.

I do use multiple IDEs, which often use Ctrl-Space for auto-completion, but so far nteract was the only application I encountered that did not support changing this.
